### PR TITLE
Multipart upload support 

### PR DESCRIPTION
**Pull request by wookie41**
Akubra now supports S3 multi part upload. It picks the backend from the active backends ring using consitent hash, sends the multi part requests to it, and when the multi part is done, Akubra writes to sync log to synchronize the rest of the backends for all clusters in a shard.

### CODE REVIEW

File: httphandler/multipart_handler.go [hunk 1]
```diff
@@ -0,0 +1,138 @@
...
+	if xmlParsingError != nil {
+
+		log.Debugf(
+			"Failed to parse body from CompleteMultipartUpload response for %s, error: %s",
+			response.Request.URL, xmlParsingError)
+
+		return false
+	}
+
+	log.Debugf("Successfully performed multipart upload to %", completeMultipartUploadResult.Location)
```
Comment (mjarco) [inline httphandler/multipart_handler.go]: %s

File: httphandler/multipart_handler.go [hunk 2]
```diff
@@ -0,0 +1,138 @@
...
+		logMsg, err := json.Marshal(syncLogMsg)
+
+		if err != nil {
+			log.Debugf("Marshall synclog error %s", err)
+			return
+		}
+
+		multiPartUploadHandler.syncLog.Println(string(logMsg))
+
+		log.Debugf("Sent a multi part uplaod migration request of object %s to backend %s", response.Request.URL.Path, backendHostName)
```
Comment (ksitak) [inline httphandler/multipart_handler.go]: upload

File: storages/multipart_picker.go [hunk 3]
```diff
@@ -0,0 +1,31 @@
+package storages
+
+import (
+	"net/http"
+
+	"github.com/allegro/akubra/log"
+)
+
+//PickRandomBackendForMultiPartUpload selects the backend that will handle the multi part upload requests and list of storages hostnames that later have to be synchronized
+func PickRandomBackendForMultiPartUpload(backends []http.RoundTripper) (multiUploadBackend *Backend, backendsHostNamesToSync []string) {
```
Comment (ksitak) [inline storages/multipart_picker.go]: it's not random

File: storages/multipart_picker_test.go [hunk 4]
```diff
@@ -0,0 +1,120 @@
...
+		Endpoint:     url.URL{},
+		Maintenance:  false,
+		Name:         "someBackend2",
+	}
+
+	backend3 := &Backend{
+		RoundTripper: nil,
+		Endpoint:     url.URL{},
+		Maintenance:  false,
+		Name:         "someBackend2",
```
Comment (ksitak) [inline storages/multipart_picker_test.go]: not 3?

File: storages/multipart_picker_test.go [hunk 5]
```diff
@@ -0,0 +1,120 @@
...
```
Comment (ksitak) [inline storages/multipart_picker_test.go]: not 3?

File: storages/multipart_round_tripper.go [hunk 6]
```diff
@@ -0,0 +1,205 @@
...
+	"github.com/allegro/akubra/httphandler"
+	"github.com/allegro/akubra/log"
+	"github.com/allegro/akubra/types"
+	"github.com/allegro/akubra/utils"
+	"github.com/serialx/hashring"
+)
+
+//MultiPartRoundTripper handles the multi part upload. If multi part upload is detected, it delegates the request
+//to handle the operation in standard fashion
+//to the backend selected using the active backends hash ring, otherwise the cluster round tripper is used
```
Comment (mjarco) [inline storages/multipart_round_tripper.go]: Style: we add extra space between // and contents

File: storages/multipart_round_tripper.go [hunk 7]
```diff
@@ -0,0 +1,208 @@
...
+
+func (multiPartRoundTripper *MultiPartRoundTripper) pickBackend(objectPath string) *Backend {
+
+	for {
+
+		backendEndpoint, _ := multiPartRoundTripper.backendsRing.GetNode(objectPath)
+		backend, ok := multiPartRoundTripper.backendsRoundTrippers[backendEndpoint]
+
+		if !ok || backend.Maintenance {
+			continue
```
Comment (mjarco) [inline storages/multipart_round_tripper.go]: hmmm infinite loop?

File: storages/multipart_round_tripper.go [hunk 8]
```diff
@@ -0,0 +1,208 @@
...
+
+	if isMultiPartUploadRequest(request) {
+
+		if multiPartRoundTripper.isNotAbleToHandleMultiUpload() {
+			return nil, errors.New("can't handle multi upload")
+		}
+
+		multiUploadBackend := multiPartRoundTripper.pickBackend(request.URL.Path)
+
+		log.Debugf("Handling multi part upload, sending %s to %s, RequestID id %s",
```
Comment (mjarco) [inline storages/multipart_round_tripper.go]: s/multi part/multipart/g

File: storages/multipart_round_tripper.go [hunk 9]
```diff
@@ -0,0 +1,208 @@
...
+
+		return false
+	}
+
+	log.Debugf("Successfully performed multipart upload to %s", completeMultipartUploadResult.Location)
+
+	return true
+}
+
+func (multiPartRoundTripper *MultiPartRoundTripper) reportCompletionToMigrator(response *http.Response, uploadedBackendName string) {
```
Comment (mjarco) [inline storages/multipart_round_tripper.go]: AFAIK response.Request.Host == uploadedBackendName, so there is no need to pass extra argument

**Pull request merged**